test/cache-tests.js

```diff
@@ -0,0 +1,258 @@
+        return cache.get('a key').should.eventually.eql({hit: true, data: new Buffer('cache1')})
+            .then((x) => {
+                subCache1.hits.should.equal(1);
+                subCache1.gets.should.equal(1);
+                subCache2.hits.should.equal(0);
+                subCache2.gets.should.equal(0);
+                return x;
+            }).then(() => {
+                Promise.all([
+                    subCache1.get('a key').should.eventually.eql({hit: true, data: new Buffer('cache1')}),
+                    subCache2.get('a key').should.eventually.eql({hit: true, data: new Buffer('cache2')})]
+                );
+            });
+    });
+});
+
+describe('On disk caches', () => {
+    basicTests(() => new OnDiskCache(newTempDir(), 10));
+    it('should evict old objects', () => {
+        const tempDir = newTempDir();
+        const cache = new OnDiskCache(tempDir, 1);
+        return cache.put('a key', 'a value', 'bob')
+            .then(() => {
+                const promises = [];
+                const oneK = "".padEnd(1024);
+                for (let i = 0; i < 1024; i++) {
+                    promises.push(cache.put(`key${i}`, oneK));
+                }
+                return Promise.all(promises);
+            })
+            .then(() => {
+                return cache.get('a key').should.eventually.contain({hit: false});
+            });
+    });
+
+    it('should handle existing data', () => {
+        const tempDir = newTempDir();
+        fs.writeFileSync(path.join(tempDir, 'abcdef'), 'this is abcdef');
+        fs.mkdirSync(path.join(tempDir, 'path'));
+        fs.writeFileSync(path.join(tempDir, 'path', 'test'), 'this is path/test');
+        const cache = new OnDiskCache(tempDir, 1);
+        return Promise.all([
+            cache.get('abcdef').should.eventually.eql({hit: true, data: new Buffer('this is abcdef')}),
+            cache.get('path/test').should.eventually.eql({hit: true, data: new Buffer('this is path/test')})]);
+    });
+
+    // MRG ideally handle the case of pre-populated stuff overflowing the size
+    // and test sorting by mtime, but that might be too tricky.
+});
+
+const S3FS = {};
+AWS.mock('S3', 'getObject', (params, callback) => {
+    params.Bucket.should.equal("test.bucket");
+    const result = S3FS[params.Key];
+    if (!result) {
+        const error = new Error("Not found");
+        error.code = "NoSuchKey";
+        callback(error);
+    } else {
+        callback(null, {Body: result});
+    }
+});
+AWS.mock('S3', 'putObject', (params, callback) => {
+    params.Bucket.should.equal("test.bucket");
+    S3FS[params.Key] = params.Body;
+    callback(null, {});
+});
+describe('S3 tests', () => {
+    basicTests(() => new S3Cache('test.bucket', 'cache', 'uk-north-1'));
+    // BE VERY CAREFUL - the below can be used with sufficient permissions to test on prod (With mocks off)...
+    // basicTests(() => new S3Cache('storage.godbolt.org', 'cache', 'us-east-1'));
```

Comment:
Risky! :)

Reply:
Right; though only if you have permissions from wherever you run the tests to access S3 :)